README
$Id: README,v 1.2 2004/04/03 15:51:26 derek Exp $

This package provides support for (initially, just reading) ISO 8211 files

http://sourceforge.net/projects/py-iso8211/

Initial code: TJ Ibbs, 17 Feb 1994
Updated: Apr 3, 2004, Derek Chen-Becker

The structure of a DDF (as held here) is:

	DDF
	 |
	[1] --- DDR (nb: sub-class of Record)
	 |	\
	 |	{ "tag" :: Field_desc }
	 |		   \
	 |		    field_controls, data_field_name,
	 |		    array_descriptor, format_controls
	 |
	[*] --- Record
	 	\
		[1] --- Leader
		 |
		[1] --- Directory
		 |	 |
		 |	[*] --- Field
		 |		\
		 |		 tag, length, data
		 |
		[1] --- Field_area


See module variables "Changes" and "Version" for the obvious things...

This code was initially written by (but no longer at this address):

	Tony J Ibbs

	Department of Geography & Topographic Science,
	University of Glasgow,
	GLASGOW  G12 8QQ
	Scotland

	Tel: (+44)141-330-6649
	Fax: (+44)141-330-4894

	Email:	tony@lsl.co.uk
	   or:	T.Ibbs@geog.gla.ac.uk

	Url: http://www.geog.gla.ac.uk/staff/res/tibbs/

Some (much) of it has been done in my own time, but some of it has been paid
for by my funders:

	Laser-Scan Ltd
	Cambridge Science Park
	Milton Road
	Cambridge CB4 4FY
	UK

	Tel: (+44)1223-420414
	Url: http://www.lsl.co.uk


Please use our sourceforge page:

http://sourceforge.net/projects/py-iso8211/

To report any bugs, feature requests, etc, or to get the latest version of
the library.

See the file INSTALL for details of how to install this package.


CHANGELOG
$Id: CHANGELOG,v 1.1 2004/04/05 17:45:50 d-rock Exp $

* 0.9.1
- Modified to work with Python 2.2 (Derek Chen-Becker)
- Changed old string.atoi() to int() (Derek Chen-Becker)
- Modified field handling to report field arrays as multiple
  fields, similar to SDTS++ (Derek Chen-Becker)


cmdline.py
#!/tools/net/bin/python

# Copyright (c) 1994, 1996, Tony J. Ibbs All rights reserved.
# Copyright (c) 2004, Derek Chen-Becker All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
# 
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#       
#     * Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#       
#     * Neither the name of py-iso8211 nor the names of its contributors
#       may be used to endorse or promote products derived from this
#       software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""commandline - a simple command line interface

	This module provides a class CommandLine which may be subclassed
	to provide command line interpretation facilities.

	This code written by:

		Tony J Ibbs

		Department of Geography & Topographic Science,
		University of Glasgow,
		GLASGOW  G12 8QQ
		Scotland

		Tel: (+44)141-330-6649
		Fax: (+44)141-330-4894

		Email:	tony@lsl.co.uk
		   or:	T.Ibbs@geog.gla.ac.uk

	and dated 10 January 1996

	NOTE - this is not specific to ISO 8211, at all, but is needed by the
	iso8211 command, so is provided in the iso8211 package for convenience.
	"""

import sys
import os
import string

# Standard stuff

FALSE = 0
TRUE  = 1

# Exceptions

class UnknownCommand(Exception):
	"Unknown command"
	pass

# The basic help string

Basic_help_string = """Command line processor.

	Commands are:

	interactive	enter a simple interactive environment.
	int		synonym for "interactive"
	debug		switch debugging on (or off)
	exit		exit the interactive environment
	help		print this message
	"""


#---------------------------------------------------------------------------------
class CommandLine:
	"""Command line context.

	This class provides the basic functionality for command line processing,
	including the ability to manage a simple sort of interactive session.

	Basic usage involves:

		- create a sub-class of CommandLine
		- add appropriate methods for the commands required
		- add the command name / method correspondences to "self.commands"
		- do
			SubClass.obey_command_line(sys.argv[1:])

	Initialisation arguments:

		None

	A CommandLine object contains:

		self.prompt		The prompt to use for interactive use
					(defaults to "Command:")

		self.separator		The separator for multiple commands on one line
					(defaults to "+")

		self.commands		A dictionary of the form:
						command name :: command method
					(defaults to entries for "int", "interactive", "help",
					"debug" and "exit")

		self.help_string	A string giving help on the basic commands already
					defined

		self.exceptions_to_ignore	A tuple of exceptions to be ignored
						(defaults to an empty tuple)

		self.exceptions_to_handle	A tuple of exceptions to be pass to self.handler
						(defaults to an empty tuple)

		self.handler		A routine to call (with arguments exception and tuple)
					to handle any exceptions that are not to be ignored,
					but do need special handling. If it raises any exceptions
					of its own, they will not be trapped.
					(defaults to None)
	"""

	def __init__(self):

		# Set up a default prompt, used when doing an interactive session

		self.prompt   = "Command: "

		# Remember if we are currently doing an interactive session

		self.in_interactive = FALSE

		# Decide on the string we use to separate multiple commands
		# on the same line - a plus makes a reasonable default, since
		# it doesn't get clobbered by the Unix shell

		self.separator = '+'

		# Define our basic commands

		self.commands = { "interactive" : self.interactive,
				  "int"		: self.interactive,
				  "help"	: self.help,
				  "debug"	: self.debug,
				  "exit"	: self.exit }

		# Define the basic help string

		self.help_string = Basic_help_string

		# We have a tuple of exceptions which should be ignored
		# when obeying a command line - these just print out
		# the basic exception type and value, which allows an
		# interactive session to continue neatly over expected
		# exceptions. Note that EOFError and UnknownCommand are
		# handled separately

		self.exceptions_to_ignore = ()	# no such exceptions, by default

		# We don't have a special exception handler

		self.exceptions_to_handle = ()	# no such exceptions, by default
		self.handler = None

		# And we're not (by default) debugging

		self.debugging = FALSE


	def __repr__(self):
		return "Command line interface"


	def exit(self,cmd,args):
		"""Request to exit interactive mode - raise an EOF error."""

		raise EOFError


	def help(self,cmd,args):
		"""The basic help command - prints the basic help string.

		Called with ARGS = None if there were no arguments on the
		command line."""

		print self.help_string


	def debug(self,cmd,args):
		"""Command to switch debugging on or off."""

		if args == None or len(args) == 0:
			self.debugging = TRUE
		elif args[0] == "true" or args[0] == "t" or args[0] == 1:
			self.debugging = TRUE
		elif args[0] == "false" or args[0] == "f" or args[0] == 0:
			self.debugging = FALSE
		else:
			print "Command is `debug [t|f|true|false|1|0]'"

		if self.debugging:
			print "Debugging on"
		else:
			print "Debugging off"


	def obey_command_line(self,command_line):
		"""Obey a command line.

		The COMMAND_LINE is a list of `words', so that this method
		may be called with "sys.argv[1:]" as a sensible argument.

		Note that a command line may be composed of multiple commands,
		separated by the separator string (which defaults to "+").
		Separators must be delimited by spaces (i.e., must count as
		`words') - this is so that (for instance) a filename may safely
		contain the separator string.
		"""

		if len(command_line) == 0:
			self.help(None)
			return

		# Split the command line apart into individual commands

		list_of_commands = self.extract_commands(command_line,self.separator)

		if self.debugging:
			print "--- command line =",list_of_commands

		# And obey them

		for command in list_of_commands:

			try:
				self.obey_command(command)

			except UnknownCommand,(cmd,args):
				# Unknown command - this should already have
				# been grumbled about, so just abort this
				# command line

				break

			except KeyboardInterrupt:
			   	# Raised by typing "Control-C"
				# Just ignore this command line

				print	# The system doesn't print a newline for us
				break

			except EOFError:
			   	# Raised by typing "exit"
				# If we're interactive, just raise it again

				if self.in_interactive:
					raise EOFError
				else:
					break

			except self.exceptions_to_ignore:
				# Any exception that we should just ignore
				# - just ignore this command line after reporting

				print "%s: %s"%(sys.exc_type,sys.exc_value)
				break

			except self.exceptions_to_handle:
				# Any exception that we should handle specially
				# - just ignore this command line afterwards

				if self.handler == None:
					print "%s: %s"%(sys.exc_type,sys.exc_value)
				else:
					self.handler(sys.exc_type,sys.exc_value)
				break



	def extract_commands(self,command_line,separator):
		"""Extract individual commands from a command line.

		Given a COMMAND_LINE (list of `words'), return a list of
		individual commands, as delimited by the SEPARATOR string.

		Each individual command is a list of command words.

		For instance, given the COMMAND_LINE [ "help" "+" "fred" "2" ],
		and the SEPARATOR "+", this routine would return:

			[ [ "help" ], [ "fred", "2" ] ]
		"""

		if self.debugging:
			print "--- extract:",command_line

		# Individual command sequences are separated by the separator string...

		commands = []

		this_command = []

		for word in command_line:
			if word == separator:
				commands.append(this_command)
				this_command = []
			else:
				this_command.append(word)
		else:
			if len(this_command) > 0:
 				commands.append(this_command)

		if self.debugging:
			print "--- which produces:",commands

		return commands



	def obey_command(self,command):
		"""Obey a command.

		The COMMAND is a list of `words', where the first is the
		command name, and the rest are any arguments.
		"""

		if self.debugging:
			print "--- command =",command

		cmd  = command[0]	# the command word
		args = command[1:]	# any arguments

		# If we recognise this command, obey the relevent function,
		# otherwise, call the "unknown command" function to take
		# appropriate action

		if self.commands.has_key(cmd):
			self.commands[cmd](cmd,args)
		else:
			self.unknown_command(cmd,args)


	def unknown_command(self,cmd,args):
		"""Called to grumble about an unknown command.

		This is given as a separate method so that it can be
		overridden by child classes, if required.
		"""

		print "Unknown command: %s"%(cmd)
		print "Commands are:",

		names = self.commands.keys()
		names.sort()

		for name in names:
			print name,
		print

		# And raise an exception to show something went wrong

		raise UnknownCommand(cmd,args)


	def interactive(self,cmd,args):
		"""Start an interactive session."""

		if self.in_interactive:
			print "Already in an interactive session - ignored"
			return
		else:
			self.in_interactive = TRUE

		# Announce ourselves

		self.interactive_intro(args)

		# And loop, reading command lines and obeying them

		while 1:
			try:
				command_line = raw_input(self.prompt)

			except KeyboardInterrupt:	# i.e., "Control-C"
				print			# need a newline
				continue		# ignore it

			except EOFError:
				print "exit"
				break

			# Split it into individual words, separated by spaces

			word_list = string.split(command_line)

			# Ignore blank lines

			if len(word_list) == 0:
				continue

			# Obey the resultant command line

			try:
				self.obey_command_line(word_list)

			except EOFError:
				# Raised by typing "exit", so do so

				self.in_interactive = FALSE
				break


	def interactive_intro(self,args):
		"""Output appropriate messages at the start of an interactive session."""

		print "Interactive command environment"
		print "(use exit or `EOF' to exit the environment)"
		print

		if len(args) > 0:
			print "The rest of the command line after an `interactive' command"\
			      " is ignored"
			print "In this case, the following commands were ignored:",
			for word in args:
				print word,
			print
			print


#---------------------------------------------------------------------------------
class Example(CommandLine):
	"""An example of subclassing the command line class."""

	def __init__(self):

		# Do the command line initialisation

		CommandLine.__init__(self)

		# Define our own data

		self.had_fred = FALSE

		# Add some commands

		self.commands["fred"] = self.fred
		self.commands["args"] = self.print_args

		# And extend the help string to know about them

		self.help_string = self.help_string + """
	fred		set the Fred flag
	args <args>	print the arguments
	"""

	def __repr__(self):
		if self.had_fred:
			return "Had Fred already"
		else:
			return "Fred"


	def fred(self,args):
		"""Implement the "fred" command."""

		if self.had_fred:
			print 'Had "fred" already'
		else:
			print 'Received "fred"'
			self.had_fred = TRUE


	def print_args(self,args):
		"""Print the arguments to the "args" command."""

		print "Arguments are:",args


#---------------------------------------------------------------------------------
def main():
	"""Main routine, for use in interactive use of this file as a command.

	Uses the "Example" command line subclass, to show how this may be used.
	"""

	# sys.argv[0] is the command the user typed, which we don't particularly
	# care about - let's just extract the rest of the command line...

	arg_list = sys.argv[1:]

	# Let's have a command line processor

	command_processor = Example()

	# And obey any commands

	command_processor.obey_command_line(arg_list)


#---------------------------------------------------------------------------------
# And if we're called from the command line, let's do it!

if __name__ == '__main__':
	main()


pathutils.py
#!/tools/net/bin/python

# Copyright (c) 1994, 1996, Tony J. Ibbs All rights reserved.
# Copyright (c) 2004, Derek Chen-Becker All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
# 
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#       
#     * Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#       
#     * Neither the name of py-iso8211 nor the names of its contributors
#       may be used to endorse or promote products derived from this
#       software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""pathutils.py - Miscellaneous useful path manipulation functions.

	This code written by:

		Tony J Ibbs

		Department of Geography & Topographic Science,
		University of Glasgow,
		GLASGOW  G12 8QQ
		Scotland

		Tel: (+44)141-330-6649
		Fax: (+44)141-330-4894

		Email:	tony@lsl.co.uk
		   or:	T.Ibbs@geog.gla.ac.uk

	NOTE - this is not specific to ISO 8211, at all, but is needed by the
	iso8211 command, so is provided in the iso8211 package for convenience.
"""

import string
import os
import posix

# I like true and false (integer is NOT boolean)

TRUE  = 1
FALSE = 0


#---------------------------------------------------------------------------------
def expand_path(inpath):
	"""Given a path INPATH which may include "~", ".", "..", etc, expand it out."""

	# Expand out any "~" or "~user" values

	inpath = os.path.expanduser(inpath)

	# Expand out any "$environ" values

	inpath = os.path.expandvars(inpath)

	# Obtain a list of the path elements, separated by "/" (for Unix)

	list = string.split(inpath,os.sep)

	# Trundle along them, composing a final path

	pathlist = []

	# If the first item is "." or "..", we need to work that
	# out wrt the actual current directory

	if list[0] == ".":
		# Get the current path and replace the "." with it
		cwd     = os.getcwd()
		cwdlist = string.split(cwd,os.sep)
		list    = cwdlist + list[1:]

		#print "Starts with . - gives us",list

	elif list[0] == "..":
		# Get the current path, remove the final element,
		# and replace the ".." with it
		cwd     = os.getcwd()
		cwdlist = string.split(cwd,os.sep)
		list    = cwdlist[:-1] + list[1:]

		#print "Starts with .. - gives us",list

	for item in list:
		if item == "":
			# An empty string means that we had the separator by itself
			# - i.e., on Unix we had "/"
			# We need to be careful of a path starting "/" (i.e., relative
			# to root), which will get turned into ["" ""]

			#print "found \"\" - ",

			if len(pathlist) == 0:
				#print "it's the first element"
				pathlist = [""]
			elif len(pathlist) == 1 and pathlist[0] == "":
				#print "it's the second element, after \"\""
				pathlist.append("")
			else:
				#print "ignoring it"
				pass


		elif item == ".":
			# The item was the current directory
			# We don't need to do anything

			#print "found . - ignoring it"
			pass

		elif item == "..":
			# The item was the previous directory
			# We need to `back up' one in our path list
			# (but look out for "/.." which means "/")

			#print "found ..",

			if len(pathlist) > 0:
				#print " - removing previous item"
				del pathlist[-1]

		else:
			# Otherwise, just add the element to the end

			#print "found %s"%item

			pathlist.append(item)

	# So we now have a list of path elements, which should be relative to root
	# - turn it back into a path

	#print "List is:",pathlist

	path = string.join(pathlist,os.sep)

	return path



__init__.py
# Copyright (c) 1994, 1996, Tony J. Ibbs All rights reserved.
# Copyright (c) 2004, Derek Chen-Becker All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
# 
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#       
#     * Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#       
#     * Neither the name of py-iso8211 nor the names of its contributors
#       may be used to endorse or promote products derived from this
#       software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""iso8211.__init__.py - header for iso8211 package

This package provides support for (initially, just reading) ISO 8211 files

Initial code: TJ Ibbs, 17 Feb 1994

The structure of a DDF (as held here) is:

	DDF
	 |
	[1] --- DDR (nb: sub-class of Record)
	 |	\
	 |	{ "tag" :: Field_desc }
	 |		   \
	 |		    field_controls, data_field_name,
	 |		    array_descriptor, format_controls
	 |
	[*] --- Record
	 	\
		[1] --- Leader
		 |
		[1] --- Directory
		 |	 |
		 |	[*] --- Field
		 |		\
		 |		 tag, length, data
		 |
		[1] --- Field_area


See module variables "Changes" and "Version" for the obvious things...

This code has been written by:

	Tony J Ibbs

	Department of Geography & Topographic Science,
	University of Glasgow,
	GLASGOW  G12 8QQ
	Scotland

	Tel: (+44)141-330-6649
	Fax: (+44)141-330-4894

	Email:	tony@lsl.co.uk
	   or:	T.Ibbs@geog.gla.ac.uk

	Url: http://www.geog.gla.ac.uk/staff/res/tibbs/

Some (much) of it has been done in my own time, but some of it has been paid
for by my funders:

	Laser-Scan Ltd
	Cambridge Science Park
	Milton Road
	Cambridge CB4 4FY
	UK

	Tel: (+44)1223-420414
	Url: http://www.lsl.co.uk

This means that the copyright for this probably belongs to them. However, I
regard this as being under the same copyright as Python itself (unless they
tell me otherwise), so please bear with that.


Regardless, ANY bug reports, awkward features, general naughtiness or simply
things you would like it to do will be happily received, although since this is
only a background task, they may not get acted on fast.


See the file INSTALL for details of how to install this package.
"""

Changes = """Modifications:

1994/02/17:	Development begun, with version 0.1
		(this work being done in my spare time!)
1994/03/04:	FROZEN: Version 0.1a
		This version read directly from the DFD, rather than using
		strings to hold the directory, etc. This approach was abandoned
		for coding simplicity, even though it is likely to be wanted at
		a later stage again.
1994/03/05:	FROZEN: Version 0.2
1994/05/03:	FROZEN: Version 0.2a

1996/03/24:	Version 0.2c (what happened to 0.2b?)
		This work is now being done partially in LSL time, to support
		other work with ISO 8211 data.

		Make the DDR also keep a *list* of tag names, and use this
		when "show"ing the fields in the DDR, so that they are shown
		in the order they occur in the DDR.
		Cope with an empty array descriptor (check if it is None
		before trying to "show" it).

1996/03/28:	Add the "explain" functionality, and change the exceptions
		raised to take account of it.
		Add support for "R" records.

1996/03/29:	Finish making support for "R" records work.
		Change the DDF class to initialise with no arguments,
		and add an "open" method to open a DDF (this seems neater).
		Tidy up the printing for field controls - add some dictionaries
		to help interpret the values, rather than "hard coding" them.
		Start work on deciphering format controls.

1996/04/16:	Include (import) the new module format.py, and use it to parse
		and show format controls.

1996/04/18:	Add support for writing out DFD files.
		Add a "Version" string for the module.

1996/04/22:	Improve the DFD writing:
		If there is no array descriptor, construct one from the
		data structure.
		If there are no format controls, construct some from the data type.

		Make the "show" methods for DDR parts simpler, now we have
		DFD writing available (to give a quick scan with show, and
		leave detail to the DFD)

1996/04/23:	Sort out the `merging' of labels and format statements in the DFD.
		Add checking of various things from the field controls, etc.
		Make an Array_descriptor contain a list of Simple_array_descriptors
		(unless it is a variable array).

1996/04/24:	Carry on with checking and `merging'
		Compare a field's data structure and data type against any
		array descriptor or format controls, and make sure they match.
		Start coping with the 0..x tag separately

1996/04/25:	Rename format.py to iso8211_format.py
		Split this (by now huge) source file into smaller files, to speed
		up editing.
		Finish coding the 0..x tags.
		Generate PARENT instructions from the 0..0 binary tree (at level 3
		only, of course).
		Add more checking of various things, and also make "common" errors
		(as determined by testing SDTS data!) give warnings and continue rather
		than raising exceptions.

1996/04/26:	Detect REPEAT TO END and write it out as such to the DFD.
		Change iso8211_format.py to use objects for format items, instead
		of tuples - this allows one to go back and *change* data in format
		items, such as noting they are "repeat to end". It is also rather
		better safer feeling than relying on "knowing" the contents of a
		tuple will (happen to) be such-and-such. (It's probably also slower.
		Ah well.)

1996/05/01:	Provide functionality to split a field into subfields.
		Make a crude attempt to deal with binary forms in that splitting
		- it doesn't cope with non-divisible-by-8 sizes of binary form, and
		it doesn't turn the binary data into floating, integer, or whatever
		(it returns a string containing the octets instead), and it doesn't
		necessarily get the octet order correct.

1996/05/15:	Version 0.3
		Fix some bugs found reading level 1 data.
		Fix writing out of DFD for unlabelled fields.

1996/05/16:	Fix writing out data for unlabelled fields, and for variable array data.
		Improve the support for binary forms:

			"B(size)" is read if "size" is a multiple of 8. An exception is
				  raised (and converted to a warning) otherwise.
				  MSOF order is assumed.

			"B"       is read. No account is taken of zero padding (so this will
				  also only really work for sizes that are multiples of 8).

			"Btw" and "btw" are read. The value is converted to MSOF.

		The values read are only partially parsed. The following are reported
		by "show()" (where "<XXX>" is a hexadecimal representation, "<IIII>" is
		an integer (Python doesn't provide unsigned integers) and "<RRRR>" is
		a real number):

			for "B(size)", "B", "B51w" and "b51w" (`unsigned integer'):

				"U:0x<XXXX> (<IIII>)"

			for "B52w" and "b52w" (`signed' integer):

				"I:0x<XXXX> (<IIII>)"

			for "B53w" and "b53w" (Fixed point real):

				"R:0x<XX>/0x<XX> (<II>/<II>=<RRRR>)"

			for "B54w" and "b54w" (Floating point real):

				"F:0x<XXXX>"

			for "B55w" and "b55w" (Complex):

				"C:0x<XXXX>,0x<XXXX>"

		Note that these are largely untested (with the exception of some testing
		on "B(size)" and "B") due to lack of data to test them on.
		Note that "C" style bit strings are always left as strings.

1996/08/28	Change version to 0.4
		Minor bug fix to checking of n500 field controls against explicit
		format controls (thanks Jamie for providing the data).
		Provide elaboration of the meaning of truncated escape sequences,
		both as hex characters, and as a description for the more common
		values (well, as used in S-57). These are written to the DFD as
		comments, as well as shown in the "show()" data.
		When reporting a record's index, it may be None, so use \%s, not \%d
		(so that we get None, rather than an exception).

		In iso8211:
		Introduce "Changes" and "Version" strings.
		Add the "show field <tag> [in record <index>]" command.
		Change "show at" to be "show record at", and tidy up how commands work.


1996/08/28	Translate character mode bitstrings (since this is fairly simple, and
		appears to work, just return the value as an integer, rather than as
		a string containing the bit string and the integer).
		Make binary form translation take account of MSOF and LSOF, and have a
		stab at a more correct implementation (issues of machine architecture
		are still to be thought about, though). Support signed binary integers.

1996/09/03	This directory of files is now a package. Use:

			import ni; ni.ni()
			import iso8211.iso8211

		and so on, to use it. The "iso8211" command is in the package. All that
		is needed to use the lot is:
		a) place this directory on the PYTHONPATH - for instance, if
		   ~/lib/python exists, and it on the PYTHONPATH, then it would be
		   sufficient to place this directory within ~/lib/python
		b) make the "iso8211" command available - either link to it, or place
		   this directory on the PATH
		Being a package means that the source files can now have shorter names
		(no more need to prefix them with "iso8211_", which is nicer.
		Open a DDF in binary mode (i.e., "rb", not just "r").

		In iso8211:
		Fix bug in "cd", "list", "dir" - they need to take account of our
		notion of current directory.
		Add "Local Variables" to end for [X]Emacs, to make this file be in
		Python mode when editing it with same.
		Add "pwd" command (horrible name, but Unix compliant).
		Make "cd" and "file" more robust (not likely to raise an exception!).
		Use "ni" so that all the iso8211 stuff is in one package - this means
		that it is now possible to place the package directory somewhere, and
		all that is needed to use this package and file is to place it somewhere
		that is on the PYTHONPATH. This also allows shorter names for the
		various sources files - no need to prefix "iso8211_" to them all.


1996/09/05	Add some initial, clumsy support for DEFINING formats to format.py

		In iso8211:
		Add the "read dfd" command, to make debugging the DFD reading code easier.
		(It isn't USEFUL yet, though, except possibly for testing the format of a
		DFD file - assuming my code is right!)


1996/09/13	Create __init__.py, and move the changes, todo list and version string
		to it, as well as the main __doc__ string.

		Add a command "package" to iso8211, which can be used to get at the version,
		etc., from the command line.

1996/09/30	Add imports of the `more important' classes and functions to __init__, so
		that they may be used directly from the package - e.g.:

			import ni; ni.ni()
			import iso8211
			ddf = iso8211.DDF()

		(instead of "ddf = iso8211.iso8211.DDF()")
		Create basic README and INSTALL texts for the package.
"""

Version = "0.4 1996/09/30"

Todo = """To do (the order is not fixed for these):
	- Add more comments, particularly with reference back into the standard,
	  explaining why things are done as they are.
	- Add specific documentation about exactly what is supported.
	- Provide documentation (!).
	- Make the separation into separate source files neater and more rational.
	- Check that a version 1 file only contains version 1 type data (at the
	  moment it may contain version 2 data).
	- Separate out REPEAT BIT FIELD in the DFD.
	- Check that format controls with bit fields are legitimate (linked to the
	  previous item).
	- Cope with extended character sets (and clause 7 in general).
	- Cope with binary forms fully and properly.
	- Add a "use_ddr()" method to DDF, such that it is possible to do:
		ddf1 = DDF()
		ddf1.open("thing.ddf","r")
		ddf1.write_DFD("broken.dfd")

		<edit the broken DFD, and create working.dfd>

		ddf2 = DDF()
		ddf2.read_DFD("working.dfd")
		ddr = ddf2.ddr
		ddf1.use_ddr(ddr)
	  This is intended to cope with cases where the DDR and data do not agree
	  - the user can extract the broken DDR into a DFD, edit it to match the
	  data, create a new DDR, and then insist on using *that* to read the data
	  with. I think I prefer the use of a function to just doing "ddf1.ddr = ddr",
	  but I'll need to think about it...
	  (Is this necessary? Well, yes, in the past something like this would have
	  been very useful, and I can imagine it being so again. It can be quite easy
	  to get a DDR that doesn't match the DRs, especially if they're created by
	  separate pieces of software.)
	- Rationalise the exceptions (there shouldn't REALLY be so many!).
	  Maybe use an exception object instead of a string.
	- Add support for reading and parsing DFD files, to create an appropriate DDR.
	- Add support for creating and writing DDFs.
"""

# Make the above more easily available - if you do:
#
#	import ni
#	ni.ni()		# (currently needed (in Python 1.3) to start "ni" up)
#	import iso8211
#	print  iso8211.Intro
#
# then you should get the sensible result...

Intro   = __doc__
#Changes = Changes
#Version = Version
#Todo    = Todo

# Similarly with the more common classes and methods...

from iso8211 import DDF
from iso8211 import Record
from iso8211 import Exceptions
from iso8211 import explain_by_hand
from iso8211 import debugging
from iso8211 import list


#__.DDF             = DDF		# Our main class
#__.Record          = Record		# Our main subsidiary class
#__.Exceptions      = Exceptions		# The tuple of our exceptions
#__.explain_by_hand = explain_by_hand	# The function that explains exceptions
#__.debugging       = debugging		# TRUE if we want debug information
#__.list            = list		# List DDF files in a directory
